### PR TITLE
CWAT-53275978-Club-Discussions-FE

### PR DESCRIPTION
**Pull request by iissh**

### CODE REVIEW

**Comment by anjalig21** [comment]
Could you provide a description of everything that is in this PR so we know what to test and review?

File: app/app/src/main/java/com/example/clubwat/model/Data.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.example.clubwat.model
+
+data class Data(
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/model/Data.kt:3]: Nit Pick - Not really important but I think most of the fe is in camel case and this is snake case

File: app/app/src/main/java/com/example/clubwat/model/UserX.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.example.clubwat.model
+
+data class UserX(
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/model/UserX.kt:3]: Why do we need UserX when we have a User class?

File: app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+interface DiscussionRepository {
+    suspend fun getClub(clubId: String, userId: String): NetworkResult<ClubDetails>
+
+    suspend fun getPosts(clubId: String, userId: String): NetworkResult<DiscussionData>
+
+    suspend fun sendMessage(request: SendDiscusionMessageRequest, bearer: String): NetworkResult<Any>
+}
+
+class DiscussionRepositoryImpl : DiscussionRepository {
+    override suspend fun getClub(clubId: String, userId: String): NetworkResult<ClubDetails> {
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt:23]: There is a similar function in ClubDetailsView. Should we put the function in a different file or just keep it in both places?

File: app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+        return NetworkResult.Error(
+            message = "API response error",
+        )
+    }
+
+    override suspend fun getPosts(clubId: String, userId: String): NetworkResult<DiscussionData> {
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt:49]: would getMessages be a better name here? not sure what posts refers to.

File: app/app/src/main/java/com/example/clubwat/viewmodels/ClubDiscussionViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    fetchUpdatedPosts(clubId) // try again after 5 seconds
+                }
+            }
+        }
+    }
+
+    private fun startPolling(clubId: String) {
+        viewModelScope.launch {
+            while (true) { // ONLY IN VM SCOPE
+                delay(60000) // delay 60s then refetch
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/viewmodels/ClubDiscussionViewModel.kt:52]: Is it possible to make the delay less? Why was 60s chosen?

File: app/app/src/main/java/com/example/clubwat/viewmodels/HomeViewModel.kt [hunk 6]
```diff
@@ -37,25 +37,4 @@ class HomeViewModel(private val userRepository: UserRepository) : ViewModel() {
             }
         }
     }
-
-    fun getAllEvents() {
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/viewmodels/HomeViewModel.kt:41]: why was this removed?

File: app/app/src/main/java/com/example/clubwat/views/HomeView.kt [hunk 7]
```diff
@@ -146,7 +146,7 @@ fun YourClubsContent(viewModel: HomeViewModel, navController: NavController) {
 @Composable
 fun YourEventsContent(viewModel: HomeViewModel, navController: NavController) {
     LaunchedEffect(key1 = true) {
-        viewModel.getAllEvents()
+        viewModel.getAllClubs()
```
Comment (Kggupta) [inline app/app/src/main/java/com/example/clubwat/views/HomeView.kt:149]: switch back to getAllEvents + add the getAllEvents back to the other thing

File: app/app/src/main/java/com/example/clubwat/views/HomeView.kt [hunk 8]
```diff
@@ -146,7 +146,7 @@ fun YourClubsContent(viewModel: HomeViewModel, navController: NavController) {
 @Composable
 fun YourEventsContent(viewModel: HomeViewModel, navController: NavController) {
     LaunchedEffect(key1 = true) {
-        viewModel.getAllEvents()
```
Comment (anjalig21) [inline app/app/src/main/java/com/example/clubwat/views/HomeView.kt:149]: should be getAllEvents

**Comment by Kggupta** [comment]
Add tasks for:
- Fixing bug when clicking the message text box, it scrolls up the club discussion header, this shouldnt happen the header should be fixed at the top. Do this in this PR if it's a small fix.
- Deleting messages (also account for club admin/regular member)

File: app/app/src/main/java/com/example/clubwat/model/Data.kt (hunk 1 above)
Comment (Kggupta) [reply app/app/src/main/java/com/example/clubwat/model/Data.kt:3]: btw, @SerializedName("") needed to be able to switch these to camel case.

File: app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            e.printStackTrace()
+        }
+        return NetworkResult.Error(
+            message = "API response error",
+        )
+    }
+
+    override suspend fun sendMessage(request: SendDiscusionMessageRequest, bearer: String): NetworkResult<Any> {
+        try {
+            val obj = URL("http://10.0.2.2:3000/api/v1/club/discussion/${request.club_id}")
```
Comment (Kggupta) [inline app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt:77]: use the env variable that other endpoint calls use instead of hardcoded url

File: app/app/src/main/java/com/example/clubwat/viewmodels/ClubDiscussionViewModel.kt (hunk 5 above)
Comment (Kggupta) [reply app/app/src/main/java/com/example/clubwat/viewmodels/ClubDiscussionViewModel.kt:52]: polling too often causes too many connections to the db stops other queries from working

**Comment by anjalig21** [comment]
heyyy

> Add tasks for:
> 
> * Fixing bug when clicking the message text box, it scrolls up the club discussion header, this shouldnt happen the header should be fixed at the top. Do this in this PR if it's a small fix.
> * Deleting messages (also account for club admin/regular member)

Should we also add time to when the user sends a message? Right now, I can only see the time for when others send messages. Also, can we format the time so it isn't 24 hours?

File: app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt (hunk 3 above)
Comment (iissh) [reply app/app/src/main/java/com/example/clubwat/repository/DiscussionRepository.kt:23]: The previous same network call is in a viewmodel, which I can't reuse in this screen. So these network calls are now in a repository that can be reused by any viewmodel.

**Comment by Kggupta** [comment]
approving, created tasks for bug fixes mentioned

**Review by Kggupta — APPROVED** [review]
